File: DotNet/CesiumLanguageWriter/Advanced/CesiumPropertyWriter.cs
using System;
using JetBrains.Annotations;

namespace CesiumLanguageWriter.Advanced
{
    /// <summary>
    /// A CZML writer for a property.  The property may be defined over a
    /// single interval or over multiple intervals.
    /// </summary>
    /// <typeparam name="TDerived">The type of the class derived from this one.</typeparam>
    public abstract class CesiumPropertyWriter<TDerived> : CesiumElementWriter, ICesiumPropertyWriter
        where TDerived : CesiumPropertyWriter<TDerived>
    {
        /// <summary>
        /// Initializes a new instance.
        /// </summary>
        /// <param name="propertyName">The name of the property.</param>
        protected CesiumPropertyWriter([NotNull] string propertyName)
            : this()
        {
            if (propertyName == null)
                throw new ArgumentNullException("propertyName");

            m_propertyName = propertyName;
        }

        /// <summary>
        /// Initializes a new instance as a copy of an existing instance.
        /// </summary>
        /// <param name="existingInstance">The existing instance to copy.</param>
        protected CesiumPropertyWriter([NotNull] CesiumPropertyWriter<TDerived> existingInstance)
            : this()
        {
            if (existingInstance == null)
                throw new ArgumentNullException("existingInstance");

            m_propertyName = existingInstance.m_propertyName;
        }

        private CesiumPropertyWriter()
        {
            m_multipleIntervals = new Lazy<CesiumIntervalListWriter<TDerived>>(CreateIntervalListWriter, false);
            m_interval = new Lazy<TDerived>(CopyForInterval, false);
        }

        /// <summary>
        /// Copies this instance and returns the copy.
        /// </summary>
        /// <returns>The copy.</returns>
        [NotNull]
        public abstract TDerived Clone();

        [NotNull]
        private CesiumIntervalListWriter<TDerived> CreateIntervalListWriter()
        {
            return new CesiumIntervalListWriter<TDerived>((TDerived)this);
        }

        /// <summary>
        /// Gets the name of the property written by this instance.
        /// </summary>
        [NotNull]
        public string PropertyName
        {
            get { return m_propertyName; }
        }

        /// <summary>
        /// Gets a value indicating whether this instance represents an open interval.
        /// </summary>
        public bool IsInterval
        {
            get { return m_elementType == ElementType.Interval || m_elementType == ElementType.PropertyConvertedToInterval; }
        }

        /// <summary>
        /// Gets a writer for intervals of this property.  The returned instance must be opened by calling
        /// the <see cref="ICesiumElementWriter.Open"/> method before it can be used for writing.  Consider
        /// calling the <see cref="OpenInterval(JulianDate,JulianDate)"/> or <see cref="OpenMultipleIntervals"/> method, which will automatically
        /// open the writer, instead of accessing this property directly.
        /// </summary>
        public TDerived IntervalWriter
        {
            get { return m_interval.Value; }
        }

        /// <summary>
        /// Gets or sets a value indicating whether this instance should always open an interval.
        /// </summary>
        public bool ForceInterval { get; set; }

        /// <summary>
        /// Opens a writer that is used to write information about this property for a single interval.
        /// </summary>
        /// <returns>The writer.</returns>
        [NotNull]
        public TDerived OpenInterval()
        {
            return OpenAndReturn(m_interval.Value);
        }

        /// <summary>
        /// Opens a writer that is used to write information about this property for a single interval.
        /// </summary>
        /// <param name="start">The start of the interval of time covered by this interval element.</param>
        /// <param name="stop">The end of the interval of time covered by this interval element.</param>
        /// <returns>The writer.</returns>
        [NotNull]
        public TDerived OpenInterval(JulianDate start, JulianDate stop)
        {
            TDerived result = OpenAndReturn(m_interval.Value);
            result.WriteInterval(start, stop);
            return result;
        }

        /// <summary>
        /// Opens a writer that is used to write information about this property for multiple discrete intervals.
        /// </summary>
        /// <returns>The writer.</returns>
        [NotNull]
        public CesiumIntervalListWriter<TDerived> OpenMultipleIntervals()
        {
            return OpenAndReturn(m_multipleIntervals.Value);
        }

        /// <summary>
        /// Writes the actual interval of time covered by this CZML interval.
        /// </summary>
        /// <param name="start">The first date of the interval.</param>
        /// <param name="stop">The last date of the interval.</param>
        public void WriteInterval(JulianDate start, JulianDate stop)
        {
            WriteInterval(new TimeInterval(start, stop));
        }

        /// <summary>
        /// Writes the actual interval of time covered by this CZML interval.
        /// </summary>
        /// <param name="interval">The interval.</param>
        public void WriteInterval([NotNull] TimeInterval interval)
        {
            if (interval == null)
                throw new ArgumentNullException("interval");

            OpenIntervalIfNecessary();
            Output.WritePropertyName("interval");
            Output.WriteValue(CesiumFormattingHelper.ToIso8601Interval(interval.Start, interval.Stop, Output.PrettyFormatting ? Iso8601Format.Extended : Iso8601Format.Compact));
        }

        /// <inheritdoc />
        protected sealed override void OnOpen()
        {
            if (m_elementType == ElementType.Interval || m_elementType == ElementType.PropertyConvertedToInterval)
                Output.WriteStartObject();
            else
                Output.WritePropertyName(m_propertyName);
        }

        /// <inheritdoc />
        protected sealed override void OnClose()
        {
            if (m_elementType == ElementType.Interval || m_elementType == ElementType.PropertyConvertedToInterval)
            {
                Output.WriteEndObject();
                if (m_elementType == ElementType.PropertyConvertedToInterval)
                    m_elementType = ElementType.Property;
            }
        }

        /// <summary>
        /// Opens an interval for this property if one is not already open.
        /// </summary>
        protected void OpenIntervalIfNecessary()
        {
            if (m_elementType == ElementType.Property)
            {
                m_elementType = ElementType.PropertyConvertedToInterval;
                OnOpen();
            }
        }

        [NotNull]
        private TDerived CopyForInterval()
        {
            TDerived result = Clone();

            CesiumPropertyWriter<TDerived> cesiumPropertyWriter = result;
            cesiumPropertyWriter.m_elementType = ElementType.Interval;

            return result;
        }

        ICesiumPropertyWriter ICesiumPropertyWriter.OpenInterval()
        {
            return OpenInterval();
        }

        ICesiumIntervalListWriter ICesiumPropertyWriter.OpenMultipleIntervals()
        {
            return OpenMultipleIntervals();
        }

        ICesiumPropertyWriter ICesiumPropertyWriter.IntervalWriter
        {
            get { return IntervalWriter; }
        }

        [NotNull]
        private readonly string m_propertyName;
        [NotNull]
        private readonly Lazy<CesiumIntervalListWriter<TDerived>> m_multipleIntervals;
        [NotNull]
        private readonly Lazy<TDerived> m_interval;
        private ElementType m_elementType = ElementType.Property;

        private enum ElementType
        {
            Property,
            Interval,
            PropertyConvertedToInterval
        }
    }
}

File: Java/CesiumLanguageWriter/translatedSrc/cesiumlanguagewriter/advanced/CesiumPropertyWriter.java
package cesiumlanguagewriter.advanced;


import agi.foundation.compatibility.*;
import agi.foundation.compatibility.ArgumentNullException;
import agi.foundation.compatibility.Enumeration;
import agi.foundation.compatibility.Func1;
import agi.foundation.compatibility.Lazy;
import cesiumlanguagewriter.*;
import javax.annotation.Nonnull;
import javax.annotation.Nullable;

/**
 * A CZML writer for a property.  The property may be defined over a
 single interval or over multiple intervals.
 * @param <TDerived> The type of the class derived from this one.
 */
@SuppressWarnings( {
        "unused",
        "deprecation",
        "serial"
})
public abstract class CesiumPropertyWriter<TDerived extends CesiumPropertyWriter<TDerived>> extends CesiumElementWriter implements ICesiumPropertyWriter {
    /**
    * Initializes a new instance.
    * @param propertyName The name of the property.
    */
    protected CesiumPropertyWriter(@Nonnull String propertyName) {
        this();
        if (propertyName == null) {
            throw new ArgumentNullException("propertyName");
        }
        m_propertyName = propertyName;
    }

    /**
    * Initializes a new instance as a copy of an existing instance.
    * @param existingInstance The existing instance to copy.
    */
    protected CesiumPropertyWriter(@Nonnull CesiumPropertyWriter<TDerived> existingInstance) {
        this();
        if (existingInstance == null) {
            throw new ArgumentNullException("existingInstance");
        }
        m_propertyName = existingInstance.m_propertyName;
    }

    private CesiumPropertyWriter() {
        m_multipleIntervals = new Lazy<cesiumlanguagewriter.CesiumIntervalListWriter<TDerived>>(new Func1<cesiumlanguagewriter.CesiumIntervalListWriter<TDerived>>(this, "createIntervalListWriter") {
            public cesiumlanguagewriter.CesiumIntervalListWriter<TDerived> invoke() {
                return createIntervalListWriter();
            }
        }, false);
        m_interval = new Lazy<TDerived>(new Func1<TDerived>(this, "copyForInterval") {
            public TDerived invoke() {
                return copyForInterval();
            }
        }, false);
    }

    /**
    * Copies this instance and returns the copy.
    * @return The copy.
    */
    @Nonnull
    public abstract TDerived clone();

    @Nonnull
    private final CesiumIntervalListWriter<TDerived> createIntervalListWriter() {
        return new CesiumIntervalListWriter<TDerived>((TDerived) this);
    }

    /**
    * Gets the name of the property written by this instance.
    */
    @Nonnull
    public final String getPropertyName() {
        return m_propertyName;
    }

    /**
    * Gets a value indicating whether this instance represents an open interval.
    */
    public final boolean getIsInterval() {
        return m_elementType == ElementType.INTERVAL || m_elementType == ElementType.PROPERTY_CONVERTED_TO_INTERVAL;
    }

    /**
    * Gets a writer for intervals of this property.  The returned instance must be opened by calling
    the {@link ICesiumElementWriter#open} method before it can be used for writing.  Consider
    calling the {@link #openInterval(JulianDate,JulianDate)} or {@link #openMultipleIntervals} method, which will automatically
    open the writer, instead of accessing this property directly.
    */
    public final TDerived getIntervalWriter() {
        return m_interval.getValue();
    }

    /**
    * Gets a value indicating whether this instance should always open an interval.
    */
    public final boolean getForceInterval() {
        return backingField$ForceInterval;
    }

    /**
    * Sets a value indicating whether this instance should always open an interval.
    */
    public final void setForceInterval(boolean value) {
        backingField$ForceInterval = value;
    }

    /**
    * Opens a writer that is used to write information about this property for a single interval.
    * @return The writer.
    */
    @Nonnull
    public final TDerived openInterval() {
        return openAndReturn(m_interval.getValue());
    }

    /**
    * Opens a writer that is used to write information about this property for a single interval.
    * @param start The start of the interval of time covered by this interval element.
    * @param stop The end of the interval of time covered by this interval element.
    * @return The writer.
    */
    @Nonnull
    public final TDerived openInterval(@Nonnull JulianDate start, @Nonnull JulianDate stop) {
        TDerived result = openAndReturn(m_interval.getValue());
        result.writeInterval(start, stop);
        return result;
    }

    /**
    * Opens a writer that is used to write information about this property for multiple discrete intervals.
    * @return The writer.
    */
    @Nonnull
    public final CesiumIntervalListWriter<TDerived> openMultipleIntervals() {
        return openAndReturn(m_multipleIntervals.getValue());
    }

    /**
    * Writes the actual interval of time covered by this CZML interval.
    * @param start The first date of the interval.
    * @param stop The last date of the interval.
    */
    public final void writeInterval(@Nonnull JulianDate start, @Nonnull JulianDate stop) {
        writeInterval(new TimeInterval(start, stop));
    }

    /**
    * Writes the actual interval of time covered by this CZML interval.
    * @param interval The interval.
    */
    public final void writeInterval(@Nonnull TimeInterval interval) {
        if (interval == null) {
            throw new ArgumentNullException("interval");
        }
        openIntervalIfNecessary();
        getOutput().writePropertyName("interval");
        getOutput().writeValue(CesiumFormattingHelper.toIso8601Interval(interval.getStart(), interval.getStop(), getOutput().getPrettyFormatting() ? Iso8601Format.EXTENDED : Iso8601Format.COMPACT));
    }

    /**
    * When overridden in a derived class, writes content to the stream immediately after opening the writer on it.
    */
    @Override
    protected final void onOpen() {
        if (m_elementType == ElementType.INTERVAL || m_elementType == ElementType.PROPERTY_CONVERTED_TO_INTERVAL) {
            getOutput().writeStartObject();
        } else {
            getOutput().writePropertyName(m_propertyName);
        }
    }

    /**
    * When overridden in a derived class, writes content to the stream immediately before closing the writer on it.
    */
    @Override
    protected final void onClose() {
        if (m_elementType == ElementType.INTERVAL || m_elementType == ElementType.PROPERTY_CONVERTED_TO_INTERVAL) {
            getOutput().writeEndObject();
            if (m_elementType == ElementType.PROPERTY_CONVERTED_TO_INTERVAL) {
                m_elementType = ElementType.PROPERTY;
            }
        }
    }

    /**
    * Opens an interval for this property if one is not already open.
    */
    protected final void openIntervalIfNecessary() {
        if (m_elementType == ElementType.PROPERTY) {
            m_elementType = ElementType.PROPERTY_CONVERTED_TO_INTERVAL;
            onOpen();
        }
    }

    @Nonnull
    private final TDerived copyForInterval() {
        TDerived result = clone();
        CesiumPropertyWriter<TDerived> cesiumPropertyWriter = result;
        cesiumPropertyWriter.m_elementType = ElementType.INTERVAL;
        return result;
    }

    @Nonnull
    private String m_propertyName;
    @Nonnull
    private Lazy<CesiumIntervalListWriter<TDerived>> m_multipleIntervals;
    @Nonnull
    private Lazy<TDerived> m_interval;
    @Nonnull
    private ElementType m_elementType = ElementType.PROPERTY;

    private static enum ElementType implements Enumeration {
        PROPERTY(0),
        INTERVAL(1),
        PROPERTY_CONVERTED_TO_INTERVAL(2);
        private final int value;

        ElementType(int value) {
            this.value = value;
        }

        /**
        * Get the numeric value associated with this enum constant.
        * @return A numeric value.
        */
        @Override
        public int getValue() {
            return value;
        }

        /**
        * Get the enum constant that is associated with the given numeric value.
        * @return The enum constant associated with value.
        * @param value a numeric value.
        */
        @Nonnull
        public static ElementType getFromValue(int value) {
            switch (value) {
            case 0:
                return PROPERTY;
            case 1:
                return INTERVAL;
            case 2:
                return PROPERTY_CONVERTED_TO_INTERVAL;
            default:
                throw new IllegalArgumentException("Undefined enum value.");
            }
        }

        /**
        * Get the enum constant that is considered to be the default.
        * @return The default enum constant.
        */
        @Nonnull
        public static ElementType getDefault() {
            return PROPERTY;
        }
    }

    private boolean backingField$ForceInterval;
}